### PR TITLE
[SPARK-32919] [SHUFFLE] Driver side changes for coordinating push based shuffle by selecting external shuffle services for merging partitions

### PR DESCRIPTION
**Pull request by venkata91**
### What changes were proposed in this pull request?

Driver side changes for coordinating push based shuffle by selecting external shuffle services for merging partitions

### Why are the changes needed?

Yes, there is a new API proposed for this use case. We can have a pluggable mechanism for selecting external shuffle services for merge process

### Does this PR introduce _any_ user-facing change?

Yes, user facing config to enable push based shuffle

### How was this patch tested?

Added Unit tests

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/internal/config/package.scala [hunk 1]
```diff
@@ -1927,4 +1927,56 @@ package object config {
...
       .booleanConf
       .createWithDefault(false)
+
+  private[spark] val PUSH_BASED_SHUFFLE_ENABLED =
+    ConfigBuilder("spark.shuffle.push.based.enabled")
+      .doc("Set to 'true' to enable push based shuffle")
+      .booleanConf
+      .createWithDefault(false)
+
+  private[spark] val PUSH_BASED_SHUFFLE_MERGE_RESULTS_TIMEOUT =
```
Comment (otterc) [inline core/src/main/scala/org/apache/spark/internal/config/package.scala]: Should not be part of this change

File: core/src/main/scala/org/apache/spark/internal/config/package.scala [hunk 2]
```diff
@@ -1927,4 +1927,56 @@ package object config {
...
+  private[spark] val PUSH_BASED_SHUFFLE_MERGE_FINALIZE_TIMEOUT =
+    ConfigBuilder("spark.shuffle.push.based.merge.finalize.timeout")
+      .doc("Specify the amount of time DAGScheduler waits after all mappers finish for " +
+        "a given shuffle map stage before it starts sending merge finalize requests to " +
+        "remote shuffle services. This allows the shuffle services some extra time to " +
+        "merge as many blocks as possible.")
+      .stringConf
+      .createWithDefault("10s")
+
+  private[spark] val PUSH_BASED_SHUFFLE_PUSHER_THREADS =
```
Comment (otterc) [inline core/src/main/scala/org/apache/spark/internal/config/package.scala]: Should not be part of this change

File: core/src/main/scala/org/apache/spark/internal/config/package.scala [hunk 3]
```diff
@@ -1927,4 +1927,56 @@ package object config {
...
+
+  private[spark] val PUSH_BASED_SHUFFLE_MERGE_RESULTS_TIMEOUT =
+    ConfigBuilder("spark.shuffle.push.based.merge.results.timeout")
+      .doc("Specify the max amount of time DAGScheduler waits for the merge results from " +
+        "all remote shuffle services for a given shuffle. DAGScheduler will start to submit " +
+        "following stages if not all results are received within the timeout.")
+      .stringConf
+      .createWithDefault("10s")
+
+  private[spark] val PUSH_BASED_SHUFFLE_MERGE_FINALIZE_TIMEOUT =
```
Comment (otterc) [inline core/src/main/scala/org/apache/spark/internal/config/package.scala]: Should not be part of this change

**Review by otterc — COMMENTED** [review]
Some of these configs are not related to these change

**Comment by venkata91** [comment]
> Some of these configs are not related to these change

Fixed it.

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala [hunk 4]
```diff
@@ -1252,6 +1254,38 @@ private[spark] class DAGScheduler(
...
   }
 
+  /**
+   * If push based shuffle is enabled, set the shuffle services to be used for the given
+   * shuffle map stage. The list of shuffle services is determined based on the list of
+   * active executors tracked by block manager master at the start of the stage.
+   */
+  private def prepareShuffleServicesForShuffleMapStage(stage: ShuffleMapStage) {
+    // TODO There are cases where reattempt of the stage could still have
+    // TODO push based shuffle enabled. This can be thought through further in the future.
```
Comment (Victsm) [inline core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala]: Not sure we should include this TODO.

File: core/src/main/scala/org/apache/spark/scheduler/ShuffleMapStage.scala [hunk 5]
```diff
@@ -88,6 +88,18 @@ private[spark] class ShuffleMapStage(
    */
   def isAvailable: Boolean = numAvailableOutputs == numPartitions
 
+  /**
+   * Retruns true if push based shuffle is disabled for this stage, or if the shuffle merge for
+   * this stage is finalized, i.e. the shuffle merge results for all partitions are available.
+   */
+  def isMergeFinalized: Boolean = {
```
Comment (Victsm) [inline core/src/main/scala/org/apache/spark/scheduler/ShuffleMapStage.scala]: This API is not used in this patch. Should we include it?
Merge finalized is not used at the beginning of a shuffle map stage.

File: core/src/main/scala/org/apache/spark/util/Utils.scala [hunk 6]
```diff
@@ -2541,6 +2541,16 @@ private[spark] object Utils extends Logging {
...
   }
 
+  /**
+   * Push based shuffle can only be enabled when external shuffle service is enabled.
+   * In the initial version, we cannot support pushed based shuffle and adaptive execution
+   * at the same time. Will improve this in a later version.
+   */
+  def isPushBasedShuffleEnabled(conf: SparkConf): Boolean = {
+    conf.get(PUSH_BASED_SHUFFLE_ENABLED) && conf.get(SHUFFLE_SERVICE_ENABLED) &&
+      !conf.getBoolean("spark.sql.adaptive.enabled", false)
```
Comment (Victsm) [inline core/src/main/scala/org/apache/spark/util/Utils.scala]: AQE no longer affects push based shuffle.

File: core/src/test/resources/META-INF/services/org.apache.spark.scheduler.ExternalClusterManager [hunk 7]
```diff
@@ -1,3 +1,4 @@
 org.apache.spark.scheduler.DummyExternalClusterManager
 org.apache.spark.scheduler.MockExternalClusterManager
 org.apache.spark.scheduler.CSMockExternalClusterManager
+org.apache.spark.scheduler.PushBasedClusterManager
```
Comment (Victsm) [inline core/src/test/resources/META-INF/services/org.apache.spark.scheduler.ExternalClusterManager]: PushBasedClusterManager is not introduced in this PR.

File: core/src/main/scala/org/apache/spark/Dependency.scala [hunk 8]
```diff
@@ -95,6 +97,41 @@ class ShuffleDependency[K: ClassTag, V: ClassTag, C: ClassTag](
...
+   * Stores the location of the list of chosen external shuffle services for handling the
+   * shuffle merge requests from mappers in this shuffle map stage.
+   */
+  private[this] var _mergerLocs: Seq[BlockManagerId] = Nil
+
+  /**
+   * Stores the information about whether the shuffle merge is finalized for the shuffle map stage
+   * associated with this shuffle dependency
+   */
+  private[this] var _shuffleMergedFinalized: Boolean = false
```
Comment (Victsm) [inline core/src/main/scala/org/apache/spark/Dependency.scala]: Is it necessary to introduce shuffleMergeFinalized in this PR?

File: core/src/main/scala/org/apache/spark/util/Utils.scala (hunk 6 above)
Comment (otterc) [reply core/src/main/scala/org/apache/spark/util/Utils.scala]: @venkata91 You need to cherry-pick this `60a8750997` from magnet-upstream

File: core/src/main/scala/org/apache/spark/shuffle/ShuffleMergerDiscovery.scala [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+import org.apache.spark.annotation.{Experimental, Since}
+import org.apache.spark.storage.BlockManagerId
+
+/**
+ * :: Experimental ::
+ * Pluggable interface to get merger locations for Push based shuffle.
+ */
+@Experimental
+@Since("3.1.0")
+trait ShuffleMergerDiscovery {
```
Comment (Victsm) [inline core/src/main/scala/org/apache/spark/shuffle/ShuffleMergerDiscovery.scala]: With pluggable API, we should expose configurations to set the implementation for this API.

File: core/src/main/scala/org/apache/spark/internal/config/package.scala [hunk 10]
```diff
@@ -1927,4 +1927,31 @@ package object config {
...
       .booleanConf
       .createWithDefault(false)
+
+  private[spark] val PUSH_BASED_SHUFFLE_ENABLED =
+    ConfigBuilder("spark.shuffle.push.based.enabled")
+      .doc("Set to 'true' to enable push based shuffle")
+      .booleanConf
+      .createWithDefault(false)
+
+  private[spark] val MAX_MERGER_LOCATIONS_CACHED =
```
Comment (Victsm) [inline core/src/main/scala/org/apache/spark/internal/config/package.scala:1937]: This configuration is not used?
Seems that you didn't add `maxRetainedMergerLocations` in `BlockManagerMasterEndpoint`.
That should be included with this PR.

File: core/src/main/scala/org/apache/spark/storage/BlockManagerMessages.scala [hunk 11]
```diff
@@ -132,6 +132,8 @@ private[spark] object BlockManagerMessages {
   case class GetReplicateInfoForRDDBlocks(blockManagerId: BlockManagerId)
     extends ToBlockManagerMaster
 
+  case object GetAllExecutors extends ToBlockManagerMaster
```
Comment (Victsm) [inline core/src/main/scala/org/apache/spark/storage/BlockManagerMessages.scala]: I don't think we still need this if we are tracking historical executor locations for merger locations.

File: core/src/main/scala/org/apache/spark/internal/config/package.scala [hunk 12]
```diff
@@ -1927,4 +1927,31 @@ package object config {
...
+    ConfigBuilder("spark.shuffle.push.retainedMergerLocations")
+      .doc("Max number of shuffle services hosts info cached to determine the locations of" +
+        " shuffle services when pushing the blocks.")
+      .intConf
+      .createWithDefault(500)
+
+  private[spark] val MERGER_LOCATIONS_MIN_THRESHOLD_RATIO =
```
Comment (Victsm) [inline core/src/main/scala/org/apache/spark/internal/config/package.scala:1944]: This and the next configuration is also not used in this PR.
Why?

File: core/src/main/scala/org/apache/spark/shuffle/ShuffleMergerDiscovery.scala [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+
+ /**
+  * Get the list of shuffle services locations in order to push the
+  * shuffle blocks for merge
+  */
+  def getMergerLocations(numMergersNeeded: Int, hostsToFilter: Set[String]) : Seq[BlockManagerId]
```
Comment (Victsm) [inline core/src/main/scala/org/apache/spark/shuffle/ShuffleMergerDiscovery.scala]: So, I'm seeing multiple getMergerLocations API here.
One is from `SchedulerBackend`:
> def getMergerLocations(numPartitions: Int, resourceProfileId: Int): Seq[BlockManagerId]

The 2nd is from `BlockManagerMaster `:
> def getMergerLocations(numMergersNeeded: Int, hostsToFilter: Set[String]): Seq[BlockManagerId]

The 3rd is from `ShuffleMergerDiscovery`:
> def getMergerLocations(numMergersNeeded: Int, hostsToFilter: Set[String]) : Seq[BlockManagerId]

DAGScheduler first invokes `getMergerLocations` API from `SchedulerBackend`, which then invokes the `getMergerLocations` API from `BlockManagerMaster`, which then invokes the `getMergerLocations` API from `ShuffleMergerDiscovery`.
This is quite confusing, since when we use the `getMergerLocations` pluggable API, the invocation has already passed through 2 layers.
The pluggable API should be the first level of invocation, which encapsulates the entire invocation path.
`hostsToFilter` shouldn't be a parameter of the API.
It should rather be the internal state managed by the actual implementation.
In the case of the default implementation, the scheduler blacklist provides `hostsToFilter`.
If we implement customized merger coordinator, it would maintain its own state for what hosts to filter.

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala (hunk 4 above)
Comment (venkata91) [reply core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala]: How about the `Reuse merger locations for sibling stages (for eg: join cases)`?

File: core/src/main/scala/org/apache/spark/shuffle/ShuffleMergerDiscovery.scala (hunk 9 above)
Comment (venkata91) [reply core/src/main/scala/org/apache/spark/shuffle/ShuffleMergerDiscovery.scala]: Can we get more feedback from OS side as to how to design this API?

File: core/src/main/scala/org/apache/spark/Dependency.scala (hunk 8 above)
Comment (venkata91) [reply core/src/main/scala/org/apache/spark/Dependency.scala]: On a closer look, We are using `shuffleMergeFinalized` as part of the `prepareShuffleServicesForShuffleMapStage` step but without having the finalized it doesn't make sense to have that. I can wait and get @zhouyejoe changes also included once he posts the PR (https://github.com/linkedin/spark/commit/818eca4199574514a1a7a1bd0b34a1d4da94ff6e).

Comment (venkata91) [reply core/src/main/scala/org/apache/spark/Dependency.scala]: For now have removed the `shuffleMergeFinalized` check eventually we need to add this back.

File: core/src/main/scala/org/apache/spark/shuffle/ShuffleMergerDiscovery.scala (hunk 13 above)
Comment (venkata91) [reply core/src/main/scala/org/apache/spark/shuffle/ShuffleMergerDiscovery.scala]: Its added to `SchedulerBackend` initially because different cluster managers can have different ways of getting merger locations. If we don't pass `hostsToFilter` to the API then we might end up with lesser than the required merger locations then we have to make another call which is why I pass this info to the API.

Comment (venkata91) [reply core/src/main/scala/org/apache/spark/shuffle/ShuffleMergerDiscovery.scala]: After having offline discussions with @Victsm decided to leave it like this and get more feedback from OS before making more changes to it.

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala [hunk 14]
```diff
@@ -1252,6 +1254,34 @@ private[spark] class DAGScheduler(
     execCores.map(cores => properties.setProperty(EXECUTOR_CORES_LOCAL_PROPERTY, cores))
   }
 
+  /**
+   * If push based shuffle is enabled, set the shuffle services to be used for the given
+   * shuffle map stage. The list of shuffle services is determined based on the list of
+   * active executors tracked by block manager master at the start of the stage.
+   */
+  private def prepareShuffleServicesForShuffleMapStage(stage: ShuffleMapStage) {
+    if (stage.shuffleDep.shuffleMergeEnabled) {
```
Comment (otterc) [inline core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala]: Can you comment here in which cases this is going to be false?
From just this patch, it seems that this will always be either true/false for all the stages because it takes the value  push-based shuffle conf.

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala [hunk 15]
```diff
@@ -1252,6 +1254,34 @@ private[spark] class DAGScheduler(
...
+      val mergerLocs = sc.schedulerBackend.getMergerLocations(
+        stage.shuffleDep.partitioner.numPartitions, stage.resourceProfileId)
+      if (mergerLocs.nonEmpty) {
+        stage.shuffleDep.setMergerLocs(mergerLocs)
+      } else {
+        stage.shuffleDep.setShuffleMergeEnabled(false)
+      }
+    } else {
+      // Disable Shuffle merge for the retry/reuse of the same shuffle dependency
+      stage.shuffleDep.setShuffleMergeEnabled(false)
```
Comment (otterc) [inline core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala]: Why is this necessary? If it comes here, then `stage.shuffleDep.shuffleMergeEnabled` will already be `false`?

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala [hunk 16]
```diff
@@ -1252,6 +1254,34 @@ private[spark] class DAGScheduler(
...
+    }
+
+    if (stage.shuffleDep.shuffleMergeEnabled) {
```
Comment (otterc) [inline core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala]: line 1276 to 1283 are just log statements for different conditions. Can't these be combined with 1263-1274?

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala [hunk 17]
```diff
@@ -1281,6 +1311,12 @@ private[spark] class DAGScheduler(
     stage match {
       case s: ShuffleMapStage =>
         outputCommitCoordinator.stageStart(stage = s.id, maxPartitionId = s.numPartitions - 1)
+        // Only generate merger location for a given shuffle map stage once. This way, even if
+        // this stage gets retried, it would still be merging blocks using the same set of
+        // shuffle services.
+        if (pushBasedShuffleEnabled && s.shuffleDep.shuffleMergeEnabled) {
```
Comment (otterc) [inline core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala]: Is it necessary to check `pushBasedShuffleEnabled` here? If yes can add to the comment when will the first condition be false and the second one be true?

File: core/src/main/scala/org/apache/spark/internal/config/package.scala (hunk 10 above)
Comment (otterc) [inline core/src/main/scala/org/apache/spark/internal/config/package.scala:1937]: This configuration is not used anywhere in this PR? Is this needed?

File: core/src/main/scala/org/apache/spark/shuffle/ShuffleMergerDiscovery.scala (hunk 9 above)
Comment (otterc) [inline core/src/main/scala/org/apache/spark/shuffle/ShuffleMergerDiscovery.scala]: It seems like even though we have introduced this API, we are still adding similar API to `SchedulerBackend`.
What is needed from `SchedulerBackends`? Is it just a list of hosts that should be excluded for mergeLocations?
In that case, should we change the API in the `SchedulerBackends` to just retrieve that list?

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala (hunk 14 above)
Comment (venkata91) [reply core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala]: Yes, since we don't have the finalize changes this code doesn't feel right. Technically I think we also need to have `shuffleMergeFinalized` changes which cannot be added due to missing `finalize` changes. This condition if you check in our internal branch, its basically `if (stage.shuffleDep.shuffleMergeEnabled && !stage.shuffleDep.shuffleMergeFinalized)` then check for merger locations and things like that else which means already `shuffleMergeFinalized` for this dependency so disable shuffle merge for the subsequent stages. Currently we don't want to enable shuffle merge for retry/reuse cases as it needs a complete refactor to handle all the cases.

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala (hunk 15 above)
Comment (venkata91) [reply core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala]: Above explanation should explain this part. With recent changes, `prepare` step depends on `finalize` changes, I'm wondering how to fix this code to avoid introducing finalize changes or it might be easier if finalize changes goes in first including the `DAGScheduler` invocation piece of code.

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala (hunk 16 above)
Comment (venkata91) [reply core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala]: This is necessary because we are both enabling as well as disabling based on the conditions so we cannot merge these log statements into the above if/else conditions.

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala (hunk 17 above)
Comment (venkata91) [reply core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala]: This is not required, will remove this part.

File: core/src/main/scala/org/apache/spark/shuffle/ShuffleMergerDiscovery.scala (hunk 9 above)
Comment (venkata91) [reply core/src/main/scala/org/apache/spark/shuffle/ShuffleMergerDiscovery.scala]: @otterc Currently this way for other schedulers like (mesos/k8s etc) push based shuffle would be disabled and also different scheduler backends can have different implementations for fetching merger locations. Like I said above, I feel better to have more feedback from OS than doing more iterations on this API now. Min was also fine with that. Let me know if you have any concerns.

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala (hunk 14 above)
Comment (otterc) [reply core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala]: I would then suggest to remove this condition. I understand it is a hassle but we need this change to start the work on the shuffle write side. The finalization part in DAG scheduler will also involve a lot of discussion because there is a question of where the finalization part will go. It seems that this should be part of the change which is related to driver finalizing the shuffle. You can add a TODO.

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala (hunk 16 above)
Comment (otterc) [reply core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala]: Can't line 1277 go after 1267 and line 1280 go after 1269?

File: core/src/main/scala/org/apache/spark/shuffle/ShuffleMergerDiscovery.scala (hunk 9 above)
Comment (otterc) [reply core/src/main/scala/org/apache/spark/shuffle/ShuffleMergerDiscovery.scala]: @venkata91 I am fine with getting more feedback from OS. However, I don't see a value of adding this API. From your response it seems that `getMergerLocations` API should be part of the scheduler backends then we probably shouldn't be adding this API. 
cc. @Victsm @mridulm

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala (hunk 14 above)
Comment (venkata91) [reply core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala]: Like you mentioned added a TODO for now.

File: core/src/main/scala/org/apache/spark/shuffle/ShuffleMergerDiscovery.scala (hunk 9 above)
Comment (venkata91) [reply core/src/main/scala/org/apache/spark/shuffle/ShuffleMergerDiscovery.scala]: @otterc This API would serve the purpose of getting merger locations from an external implementation but different scheduler backends can choose to have their own implementation or reuse the existing implementation for getting merger locations. I feel merger locations should be abstracted away from DAGScheduler since we have multiple SchedulerBackends and having only one implementation might be too strict.

Comment (Victsm) [reply core/src/main/scala/org/apache/spark/shuffle/ShuffleMergerDiscovery.scala]: I feel that we should also include application ID and shuffle ID in the API.
This is not needed now, but might be necessary if people set out to build an external merger location discovery service.

Comment (venkata91) [reply core/src/main/scala/org/apache/spark/shuffle/ShuffleMergerDiscovery.scala]: Makes sense, then I will resolve this conversation for now.

Comment (Victsm) [reply core/src/main/scala/org/apache/spark/shuffle/ShuffleMergerDiscovery.scala]: Since this API itself at the current stage is still very lightweight, I think what might be more reasonable is to not include `ShuffleMergerDiscovery` initially in the PR.
Let's send out the PR w/o this API, and then we can raise the discussion in the community about the need of it.
I think this way is probably better for the community to review and chime-in.
Currently, adding this API into the PR could cause some confusions.

**Comment by venkata91** [comment]
@otterc @Victsm Do you think we can proceed with raising this PR against OS now? If there are any other concerns let me know I can address those before going forward. Thanks

**Comment by otterc** [comment]
I don't have any concerns. The earlier we open this against OS, the better it is for us. So please go ahead.

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala [hunk 18]
```diff
@@ -1252,6 +1254,28 @@ private[spark] class DAGScheduler(
     execCores.map(cores => properties.setProperty(EXECUTOR_CORES_LOCAL_PROPERTY, cores))
   }
 
+  /**
+   * If push based shuffle is enabled, set the shuffle services to be used for the given
+   * shuffle map stage. The list of shuffle services is determined based on the list of
+   * active executors tracked by block manager master at the start of the stage.
+   */
+  private def prepareShuffleServicesForShuffleMapStage(stage: ShuffleMapStage) {
+    // TODO: Handle stage reuse/retry cases separately as without finalize changes we cannot
```
Comment (JulySouthWind) [inline core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:1263]: Maybe, we can add a shuffle service master role to manage the shuffle service worker. App driver requests shuffle service workers from shuffle service master. The master can customize the politics to assign the workers for apps. Also, the master manages the blacklist of shuffle service worker, adding or removing shuffle workers. In addition, the master can notify worker to delete data of completed apps by heart beat.

Comment (otterc) [reply core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:1263]: @JulySouthWind I think you might have commented here by mistake. The relevant PR that is actively being worked on is https://github.com/apache/spark/pull/30164.
This was a linkedin internal PR that was created but now we will just close this.
